hat/app/org/hatdex/hat/api/controllers/devices/MachineData.scala

```diff
@@ -0,0 +1,338 @@
+            handleCreateDeviceData(
+              hatUser = hatUser,
+              machineDataCreate = deviceDataCreate,
+              namespace = namespace,
+              endpoint = endpoint,
+              skipErrors = skipErrors
+            )
+        }
+      }
+
+    }
+
+  def updateData(namespace: String): Action[DeviceDataUpdateRequest] =
+    UserAwareAction.async(parsers.json[DeviceDataUpdateRequest]) { implicit request =>
+      val deviceDataUpdate = request.body
+      val eventuallyVerdict =
+        deviceVerification.getRequestVerdict(request.headers,
+                                             request.dynamicEnvironment.hatName,
+                                             namespace,
+                                             usersService,
+                                             trustedApplicationProvider
+        )
+
+      eventuallyVerdict.flatMap { verdict =>
+        verdict match {
+          case Left(verdictError) => handleFailedRequestAssessment(verdictError)
+          case Right(DeviceVerificationSuccess.DeviceRequestSuccess(hatUser)) =>
+            handleUpdateDeviceData(hatUser, deviceDataUpdate, namespace)
+        }
+      }
+
+    // Handle these errors
+    // val deviceDataUpdate    = request.body
+    // val maybeDeviceDataInfo = refineDeviceDataUpdateRequest(deviceDataUpdate)
+    // maybeDeviceDataInfo match {
+    //   case Some(deviceDataInfo) =>
+    //     deviceValid(deviceDataInfo, "NAMESPACE").flatMap { deviceOk =>
+    //       deviceOk match {
+    //         case (Some(hatUser), Right(RequestVerified(ns))) =>
+    //           handleUpdateDeviceData(hatUser, deviceDataUpdate, ns)
+    //         case (_, Left(x)) => handleFailedRequestAssessment(x)
+    //         case (None, Right(_)) =>
+    //           logger.warn(s"UpdateDevice: Hat not found for:  ${deviceDataUpdate}")
+    //           handleFailedRequestAssessment(HatNotFound(deviceDataInfo.hatName.toString))
+    //         case (_, _) =>
+    //           logger.warn(s"UpdateDevice: Fallback Error case for:  ${deviceDataUpdate}")
+    //           handleFailedRequestAssessment(GeneralError)
+    //       }
+    //     }
+    //   case None => Future.successful(BadRequest("Missing Device Details."))
+    // }
+    }
+
+  // -- Operations
+
+  // -- Get Data
+  private def makeData(
+      namespace: String,
+      endpoint: String,
+      orderBy: Option[String],
+      ordering: Option[String],
+      skip: Option[Int],
+      take: Option[Int]
+    )(implicit db: HATPostgresProfile.api.Database): Future[Result] = {
+    val dataEndpoint = s"$namespace/$endpoint"
+    val query =
+      Seq(EndpointQuery(dataEndpoint, None, None, None))
+    val data = dataService.propertyData(
+      query,
+      orderBy,
+      ordering.contains("descending"),
+      skip.getOrElse(0),
+      take.orElse(Some(defaultRecordLimit))
+    )
+    data.map(d => Ok(Json.toJson(d)))
+  }
+
+  // -- Create Data
+  def handleCreateDeviceData(
```

Comment:
only the three methods of the controller that are invoked by Play should be public, all other helpers should be private